Refactor status-go configuration

pedropombeiro opened the pull request:
Important changes:
- We now validate the configuration on `StatusBackend.startNode()`;
- Node config fields which specified a file were removed. Fields that referenced a private key file were replaced with a hex encoding of the private key;
- Each config struct has a `Validate` method which will check things such as whether we forgot to specify required directories, genesis is missing and cannot be calculated, or if there are conflicting settings specified;
- Improved output of some test suits (e.g. some errors were not being brought to the attention of the dev);
- Removed unused `WhisperConfig.FirebaseConfig` field;
- There are 4 new config files under `static/config/cli`, `fleet-eth.beta.json` is used if the app is started without arguments. If we find it useful, we can expose a command line argument allowing to preload any of those files on demand, e.g. `-load=fleet-eth.staging,les-enabled`; 
- ~~Environment variables can be used to override configuration values. Example:~~
    ``` sh
    STATUSD_CLUSTERCONFIG_FLEET=eth.staging build/bin/statusd-c /home/pedro/Downloads/default.json -c /home/pedro/Downloads/cluster.json
    ```
    ~~`ClusterConfig.Fleet` will be set to `eth.staging` even though `default.json` specified `eth.beta`.~~

Notes to reviewers:
- Ended up not flattening the JSON since most of the problems we had were addressed (`nil` pointers and being able to override settings from command line env vars);
- Let me know if there are additional validations we can put in place in the config that I might have missed;
- Validation is done in the backend (`StartNode`) and in `params.LoadNodeConfig`. I didn't move it deeper in the call stack (e.g. `MakeNode`) since that would preclude tests from passing an empty `DataDir` for in-memory database.

New output of `statusd --help`: 
``` shell
Usage: statusd [options]
Examples:
  statusd -c ./default.json                      # run node with configuration specified in ./default.json file
  statusd -c ./default.json -c ./standalone.json # run node with configuration specified in ./default.json file, after merging ./standalone.json file
  statusd -c ./default.json -metrics             # run node with configuration specified in ./default.json file, and expose ethereum metrics with debug_metrics jsonrpc call
  statusd -c ./default.json -cli                 # run node with configuration specified in ./default.json file, and enable connection by statusd-cli on default port
  STATUSD_CLUSTERCONFIG_FLEET=eth.staging \
    statusd -c ./default.json                    # run node with configuration specified in ./default.json file, and override ClusterConfig.Fleet value

Options:
  -c value
        JSON configuration file(s). Multiple configuration files can be specified, and will be merged in occurrence order
  -cli
        Enable debugging CLI server
  -cli-port string
        CLI server's listening port (default "51515")
  -log-without-color
        Disables log colors
  -metrics
        Expose ethereum metrics with debug_metrics jsonrpc call
  -pprof
        Enable runtime profiling via pprof
  -pprof-port int
        Port for runtime profiling via pprof (default 52525)
  -sync-and-exit int
        Timeout in minutes for blockchain sync and exit, zero means no timeout unless sync is finished (default -1)
  -version
        Print version and dump configuration
```

Closes #1180

Commit:
Streamline configuration in status-go. Part of #1180

- Replace command line flags with `-c` config flag. Part of #1180
- Convert node config private keys to hex-encoded string versions.
- Remove `GenerateConfig` from library.
- Remove unused `FirebaseConfig` from library.
- Fix loading of `config/status-chain-genesis.json` in non-dev machines.

pedropombeiro commented on node/status_node.go:
```diff
@@ -186,7 +187,7 @@ func (n *StatusNode) setupRPCClient() (err error) {
 
 func (n *StatusNode) discoveryEnabled() bool {
-	return n.config != nil && (!n.config.NoDiscovery || n.config.Rendezvous) && n.config.ClusterConfig != nil
+	return n.config != nil && (!n.config.NoDiscovery || n.config.Rendezvous) && n.config.ClusterConfig.Enabled
```
Do we want to rely on `len(n.config.ClusterConfig.Bootnodes) > 0` here?

adambabik replied:
As far as I know, the discovery service may use a regular node as well if it runs a discovery protocol. Another case is that rendezvous is used. I would keep it as it is.

PombeirP commented on params/config.go:
```diff
@@ -54,33 +43,13 @@ type LightEthConfig struct {
-	// AuthorizationKeyFile file path that contains FCM authorization key
-	AuthorizationKeyFile string
+	// AuthorizationKey contains FCM authorization key
+	AuthorizationKey string
```
This doesn't seem to be used at all, or am I missing something?

adambabik commented on cmd/statusd/flags.go:
```diff
@@ -1,26 +1,26 @@
-	trimmed := strings.TrimSpace(value)
-	if len(trimmed) == 0 {
-		return ErrorEmpty
+func (f *configFlags) Set(value string) error {
```
I think you can write it without nested `if`s. Also `stat, err := os.Stat(value)` may create a new scoped `err`:
```go
stat, err := os.Stat(value)
if err != nil {
  return fmt.Errorf("invalid config file: %v", err)
}
if stat.IsDir() {
  return fmt.Errorf("config file can not be a dir")
}
*f = append(*f, value)
return nil
```

adambabik commented on cmd/statusd/main.go:
```diff
@@ -318,6 +190,24 @@ func configureStatusService(flagValue string, nodeConfig *params.NodeConfig) (*p
 	return nodeConfig, nil
 }
 
+func parseConfig(configFiles configFlags, config *params.NodeConfig) bool {
```
Returning `bool` is useful in any case?

pedropombeiro replied:
if no files are specified (false is returned), main will show the version information, since it has no config to start the app. This is a change form previous behaviour, where the app would start with defaults.

adambabik commented on cmd/statusd/main.go:
```diff
@@ -318,6 +190,24 @@ func configureStatusService(flagValue string, nodeConfig *params.NodeConfig) (*p
+	configSpecified := len(configFiles) > 0
+	if configSpecified {
+		// Merge specified configuration files, in order
+		if err := params.ReadConfigFiles(configFiles, config); err != nil {
```
`ReadConfigFiles` should handle also `nil` and empty slices, right?

pedropombeiro replied:
Not sure I understand your suggestion. Do you mean we should not check the length of `configFiles` and instead just call `ReadConfigFiles`? I decided not to allow running the app without passing at least one config file, since the alternative of having everything specified with env vars doesn't look really useful.

adambabik replied:
>  Do you mean we should not check the length of configFiles and instead just call ReadConfigFiles?

Correct. But we do have some default values when calling `NewNodeConfig`. Is this not enough to start the node?

pedropombeiro replied:
Yeah, makes sense. I'll make sure the node can be started with default config.

pedropombeiro replied:
@adambabik does this mean that we'll still want to populate the `ClusterConfig.BootNodes` with the values for `eth.beta` as in current `develop`?

adambabik replied:
Hm I would like to say no. To start it with `eth.beta`, one would need to do `./statusd -c beta-fleet.json`. However, now the question is what's the point of running `statusd` without any peers? 

So maybe you're right that it should not be possible to start `statusd` without a config but we should provide some predefined JSON configs in the repo.

adambabik commented on mailserver/mailserver.go:
```diff
@@ -172,8 +173,13 @@ func (s *WMailServer) setupRequestMessageDecryptor(config *params.WhisperConfig)
+	if config.MailServerAsymKey != "" {
+		var keyAsym *ecdsa.PrivateKey
+		var convertErr error
+		if keyAsym, convertErr = crypto.HexToECDSA(config.MailServerAsymKey); convertErr != nil {
```
```go
keyAsym, err := crypto.HexToECDSA(config.MailServerAsymKey)
if err != nil {
  return err
}
s.asymFilter = &whisper.Filter{KeyAsym: keyAsym}
```

pedropombeiro replied:
You're right, that looks better, thanks.

adambabik commented on params/cluster.go:
```diff
@@ -14,62 +14,3 @@ type Cluster struct {
 }
-
-// Consult this list with http://fleets.status.im/.
-var clusters = map[string]func() Cluster{
```
Has it been moved to a JSON file?

pedropombeiro replied:
This list will at least live in status-react through https://github.com/status-im/status-react/issues/5739, but I guess we could create one config file per cluster somewhere in the repo for ease-of-use. I had done this already but just for the `eth.test` cluster in https://github.com/status-im/status-go/blob/feature%2F1180-simple-config/.vscode/node-config.json.

pedropombeiro replied:
Another option could be for `statusd` to download the list from fleets.status.im. wdyt?

adambabik replied:
Yeah, I think we should create `config/` directory and put there some predefined values (`eth-beta.json`, `eth-staging.json`, `les-enabled.json` so I can do `-c config/eth-staging.json -c config/les-enabled.json`) in order to keep the development and testing easy.

adambabik replied:
> Another option could be for statusd to download the list from fleets.status.im. wdyt?

We can have a bot or script that would periodically check that and send a PR if the values have changed. We can't do it on the fly due to predictable builds effort. 

However, we might create an option `-fleet=https://some.address/fleet.json`but I don't see how it can be useful now.

pedropombeiro replied:
yeah, I was thinking it would be part of the `make setup` target to update files in a `.gitignored` folder. So it would guarantee that users would have the latest version of the file when they set up the environment. But the bot idea is also something I've discussed with @yenda and think we should put into something like Probot in the future.

adambabik commented on node/node.go:
```diff
@@ -113,7 +110,7 @@ func MakeNode(config *params.NodeConfig, db *leveldb.DB) (*node.Node, error) {
 
 // defaultEmbeddedNodeConfig returns default stack configuration for mobile client node
-func defaultEmbeddedNodeConfig(config *params.NodeConfig) *node.Config {
+func defaultEmbeddedNodeConfig(config *params.NodeConfig) (*node.Config, error) {
```
This is very confusing to me. Can we rename it to `newGethNodeConfig`? We use `NodeConfig` here but also `NodeConfig` for the whole `statusd`.

adambabik commented on params/config.go:
```diff
@@ -365,51 +278,47 @@ type NodeConfig struct {
-	// adjust dependent values
-	if err := nodeConfig.updateConfig(); err != nil {
+	// Add genesis configuration
+	if err := nodeConfig.updateGenesisConfig(); err != nil {
```
I think it's not necessary. I would track how it's used but there is no point to put the whole genesis into config. I would calculate it only when needed.

pedropombeiro replied:
You mean calculate it in `activateLightEthService` if we haven't received right?
What would you see happening in `ValidateNodeConfig`? Would we try to decode the JSON if not empty just to check that it is valid JSON?

pedropombeiro replied:
One thing we lose by doing that is that calculated `LightEthConfig.Genesis` is not output by `statusd -version`, so it's not longer showing the final version of the config.

adambabik replied:
I reason about it this way that only `NetworkID` is required and `Genesis` is always derived from `NetworkID`. That's why we don't need to present it when executing `statusd -version`. In `ValideNodeConfig`, we only check `NetworkID` presence or check if it belongs to a set of possible values. `Genesis` is calculated when executing `activateLightEthService`. If it's not found, we just return an error.

pedropombeiro replied:
If `Genesis` is always derived from `NetwordID`, is there any reason to keep it in `LightEthConfig`?

pedropombeiro replied:
Latest commit removes `LightEthConfig.Genesis` and just calculates it on `activateLightEthService`

adambabik commented on params/config.go:
```diff
@@ -418,37 +327,38 @@ func NewNodeConfig(dataDir, clstrCfgFile, fleet string, networkID uint64) (*Node
-		return nil, err
+	if nodeConfig.LightEthConfig.Enabled && nodeConfig.LightEthConfig.Genesis == "" {
+		// Add genesis configuration
+		if err := nodeConfig.updateGenesisConfig(); err != nil {
```
Here as well.

adambabik commented on params/config.go:
```diff
@@ -473,70 +383,176 @@ func (c *NodeConfig) Validate() error {
-		if err := validate.Struct(c.ClusterConfig); err != nil {
-			return err
-		}
+	if c.DataDir == "" {
```
Isn't it handled with `validate.Struct(c)`?

adambabik commented on params/config.go:
```diff
@@ -473,70 +383,176 @@ func (c *NodeConfig) Validate() error {
-			return err
+	if c.NodeKey != "" {
+		if _, err := crypto.HexToECDSA(c.NodeKey); err != nil {
+			return fmt.Errorf("NodeKey is invalid: %s", c.NodeKey)
```
It would be nice to put `error` here. `return fmt.Errorf("NodeKey '%s' is invalid: err", c.NodeKey, err)`

pedropombeiro replied:
Good point, missed that.

adambabik commented on params/config.go:
```diff
@@ -473,70 +383,176 @@ func (c *NodeConfig) Validate() error {
+	if !c.NoDiscovery && len(c.ClusterConfig.BootNodes) == 0 {
+		// No point in running discovery if we don't have bootnodes.
+		// In case we do have bootnodes, NoDiscovery should be true.
+		return fmt.Errorf("NoDiscovery is false, but ClusterConfig.BootNodes is empty")
```
I am not sure about it. Discovery might still work if there are peers that have it enabled.

dshulyak replied:
@adambabik can you please clarify what do you mean?
if there are no bootnodes then discovery server won't be connected to any other discovery server. unless someone will try to connect to the discovery server. but they must specifically add this peer as their own bootnode. creating normal p2p connections will not affect discovery.

adambabik replied:
I might be wrong here but I though that if a node is connected to peers that support a discovery protocol then bootnodes are not necessary required.

divan commented:
Some pre-review comments.

> Environment variables can be used to override configuration values.
I would strongly suggest not to do that. The main problem with configuration is inability to say at runtime, where the particular configuration value was obtained from – hardcoded defaults, command line, config file, config parser defaults or environment variable.

The more ways to specify config value we have, the more complexity we put into the configuration part of the program, and the harder it becomes to debug in the future. 

I had strong assumption that this config refactoring effort is aimed to simplify configuration and make it easier to reason about. Adding environment variables overrides without a really strong need for it, breaks this assumption.

---
On the new output part:
```
Examples:
  statusd -c ./default.json                      # run node with configuration specified in ./default.json file
  ...
```

This effectively adds a requirement to have config file around in order to run the `statusd`. I would say it's a downgrade in usability.
If those values are really default ones, why not hardcode them as a "default zero value for config" – i.e. config that would work for the majority of use cases for command line version. So the statusd works when users just run:
```statusd```

I would also suggest to keep most usable configuration presets hardcoded (like for Whisper or Swarm enabled, etc, based on our real-world usage of statusd).

pedropombeiro commented:
Thanks for the comments @divan.
- Regarding the command line options, that is a very minor part of this PR, and very easy to get rid of. However, I would like to better understand the reasoning for doing that. In my view, that mechanism is only for doing testing on your laptop, playing with different values in a script (e.g. playing with different PoW values, or something along those lines). It's not something meant for permanent configuration. For example, it's not something I'd see Jacob using in the Ansible configuration on top of the JSON configuration he'd pass to the nodes. So I guess I don't really see the harm in being able to override settings using environment variables.
- > I would also suggest to keep most usable configuration presets hardcoded (like for Whisper or Swarm enabled, etc, based on our real-world usage of statusd).

  This refactor certainly doesn't preclude doing that, it would be very easy to add a command line option such as `-configure=mailserver` that would load the intended defaults. Please also see [discussion above](https://github.com/status-im/status-go/pull/1183#discussion_r215949927) for some additional context. There are two concerns I have with the approach though:
    - One drawback of keeping defaults in `statusd` is that now you have cluster lists in both the status-react and status-go repos, and they can get out of sync. One suggestion I made at the beginning was that we could dynamically update the fleet list from fleets.status.im as part of `make setup`. Of course, other values could still get out of sync over type with status-react, such as the default PoW value used.
    - If defaults are embedded in `statusd`, they are not easily version-controlled, so it's hard to know what the values embedded in a certain default are. I guess you could do `statusd -configure=mailserver -version` to output the JSON and compare that. But maybe we just need to establish the use cases for this scenario. For example, I don't see Jacob wanting to rely on this approach instead of having a full JSON file.

divan commented:
> It's not something meant for permanent configuration.

Sorry for answering with a picture, but it's the most eloquent answer I can provide :)
![design-vs-user-experience](https://user-images.githubusercontent.com/880202/45356065-f5672900-b5ca-11e8-8afa-51e3bf9507e2.jpg)

There are many cases, where it'll be much easier to define variable (docker containers, docker-compose, someone just preferring env vars, 12-factor app proponents, matter of personal taste, etc). Two ways of doing the same things virtually always split the knowledge, docs, practices and people, but with configuration, it creates even more harm by making it hard to reason about "where this value came from".

One of the use cases I really like to provide is hackaton, where we have to ask 8000 devs to hack something with `status-go`. The easier the entrance barrier for them and the simpler configuration the better. I can practically guarantee that with this approach they will split into groups, where someone prefers env vars, and others - config files. And even more – those people unlikely to deeply know codebase and know all intricacies of config system, so when they'll have to debug some issues/bugs with configuration, they spend much more time by deciphering where the config value really comes from.

So, in my experience, multiple config sources create much more harm than good.

 > One drawback of keeping defaults in statusd is that now you have cluster lists in both the status-react and status-go repos, and they can get out of sync
Fair point and that's probably a separate problem to solve, but still, I think that `statusd` should work out of the box. Even user without knowing what fleets are and how to configure them, just playing around with statusd, should be able to run it.

Again, back to my hackaton example, at ETHDenver there was a crappy internet, and people were literally sending sources and binaries to each other via AirDrop. It's easy to foresee that at some point requiring a JSON file to be shipped with binary will become a hurdle. Or people who just exploring list of project to hack on, could try to run `statusd`, see that it doesn't work out of the box (asking for mysterious `default.json`) will just simply switch to another project.

The point is, we should not make statusd unusable if run as is out of the box. I think this is true for any software. Nobody likes software that doesn't work unless provided some configs or tons of command line flags. It should work by default for 80% of use cases.

pedropombeiro commented:
Nice use of the UX/design picture for env vars :-), as I said, I'm not attached to it and it can easily be removed and just as easily added back in the future should we ever change our minds.

As for running `statusd` out of the box, do you think we should just include a `cli-default.json` file the same way as `static/config/status-chain-genesis.json` containing the existing defaults at `origin/develop`?

pedropombeiro commented:
By the way, I just noticed that loading the `status-chain-genesis.json` will only work in a system with the repo at the expected location. If I use the `statusd` executable on another machine, it will fail. That doesn't seem like the desired behavior, right? cc @adambabik 

See https://github.com/status-im/status-go/blob/develop/params/config.go#L575

<blockquote><img src="https://avatars3.githubusercontent.com/u/11767950?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-go">status-im/status-go</a></strong></div><div>The Status module that consumes go-ethereum. Contribute to status-im/status-go development by creating an account on GitHub.</div></blockquote>

adambabik commented:
> By the way, I just noticed that loading the status-chain-genesis.json will only work in a system with the repo at the expected location. If I use the statusd executable on another machine, it will fail. That doesn't seem like the desired behavior, right? 

Yeah, it's not. If Rinkeby, Ropsten, Mainnet chain geneses are built into a binary, it should be as well.

Regarding the default config, it's a valid point that it should be possible to start `statusd` without `default.json`. At the same time, we should keep default values in one place. Currently, some default values are in `params` package and some are in `cmd/statusd`. 

I imagine having an interface like this:
```go
// NewConfig returns a zero-value struct with required field values.
NewConfig(datadir string, network int, any other required params?)

// NewDefaultConfig returns a config object with required field values,
// Whisper enabled and beta fleet.
NewDefaultConfig(datadir string, network int, any other required params?)
```

`NewDefaultConfig` can be used by `statusd` if there is no config provided. `NewConfig` will be used when creating a node programatically. In case of config files or JSON config, `NewConfigFromFiles(files string...) or NewConfigFromJSON(files string)` can be used.

In this case, everything is explicit and all use cases are covered and UX should not be hurt.

adambabik commented:
Regarding env variables, I am ambivalent. On the one hand, they replace flag options that we are removing in this PR, on the other hand, they introduce one more layer that might be confusing. However, we don't need to advertise this option. We can treat it as an advanced feature.

pedropombeiro commented:
Yeah, `NewNodeConfig`/`NewNodeConfigWithDefaults` is the way I'm currently approaching it, even though `NewNodeConfig` still has quite a few (safe) defaults, such as `log`, `ListenAddr`, `APIModules`, `Upstream.URL`, etc.

pedropombeiro commented:
@adambabik OK, I'll partially revert https://github.com/status-im/status-go/commit/0f5ad2545da053e3f7da1336048f831f78836b15 for genesis data.

@divan what do you think about @adambabik's proposal of leaving the env vars as a hidden advanced feature? Does that still feel like the wrong thing to do to you?

divan commented:
> what do you think about @adambabik's proposal of leaving the env vars as a hidden advanced feature? Does that still feel like the wrong thing to do to you?

I have to say yes here again. "Hiding" or "not advertising" things is as effective as ["security through obscurity"](https://en.wikipedia.org/wiki/Security_through_obscurity) – it gives a false hope that thing will not be abused, but it doesn't work in practice.

From my point of view, having a single (or at list keep to a very minimum) source of configuration is a goal – it's easy to deal with, reason about and keep track of. Leaving a few flags for most common cases/needs is OK, but should be used sparingly (and it's simpler and easier to use than env vars).

Plus, bringing another dependency with over 1K LOC for the doubtful feature that potentially introduces confusion and reduces the clarity of configuration, is just not worth it.

adambabik commented:
I don't see any relation between reading env variables and security. It's just an advanced option. It gives a choice but it does not affect security. You can't do anything more with that compared to using a config file. (Regarding security, using env vars to keep credentials is more secure than storing them in a plain text file actually.)

Having said that, let's remove reading env variables. It introduces disagreement while promising something that we are not even sure anyone will benefit from.

pedropombeiro commented:
OK, removed env vars stuff, and added a paragraph to the OP:
> - There are 4 new config files under `static/config/cli`, `fleet-eth.beta.json` is used if the app is started without arguments. If we find it useful, we can expose a command line argument allowing to preload any of those files on demand, e.g. `-load=fleet-eth.staging,les-enabled`; 

We don't have to implement that in this PR, but it'll probably be something we'll want at some point.

pedropombeiro commented:
Is there anything missing for the review to proceed?

adambabik commented on Makefile:
```diff
@@ -198,6 +198,10 @@ xgo:
 	go get github.com/karalabe/xgo
 
 setup: dep-install lint-install mock-install ##@other Prepare project for first build
+	go get -u github.com/kevinburke/go-bindata/go-bindata
```
It suggests that we add config files to the binary. Is that true?

pedropombeiro replied:
Yes, it's the only way we can run `statusd` without options, as @divan requested, or am I missing something?

adambabik commented on api/backend_test.go:
```diff
@@ -9,13 +9,14 @@ import (
 func TestBackendStartNodeConcurrently(t *testing.T) {
 	backend := NewStatusBackend()
-	config := params.NodeConfig{}
+	config, _ := utils.MakeTestNodeConfig(params.StatusChainNetworkID)
```
Any reason not to check this error?

pedropombeiro replied:
The code is simple setup that should never fail, but I guess it doesn't hurt to assert it. Will change it.

adambabik commented on node/node.go:
```diff
@@ -139,26 +137,71 @@ func defaultEmbeddedNodeConfig(config *params.NodeConfig) *node.Config {
+			return nil, err
+		}
+	default:
+		return nil, nil
```
Shouldn't it return an error if there is no genesis found?

adambabik commented on params/config.go:
```diff
@@ -364,91 +274,142 @@ type NodeConfig struct {
+		return nil, fmt.Errorf("Default config could not be loaded: %s", err)
+	}
+
+	nodeConfig.NoDiscovery = true
```
Are you sure we want to disable discovery by default?

pedropombeiro replied:
I tried to keep the existing values currently reported in `origin/develop`

adambabik commented on params/config.go:
```diff
@@ -364,91 +274,142 @@ type NodeConfig struct {
+		NetworkID:       networkID,
+		DataDir:         dataDir,
+		Version:         Version,
+		RPCEnabled:      RPCEnabledDefault,
```
`RPCEnabledDefault` is used only here. I would get rid of it and keep it `false`.

adambabik commented on params/config.go:
```diff
@@ -364,91 +274,142 @@ type NodeConfig struct {
+		DataDir:         dataDir,
+		Version:         Version,
+		RPCEnabled:      RPCEnabledDefault,
+		HTTPHost:        HTTPHost,
```
The same with this one, I don't think HTTP interface should be enabled by default.

adambabik commented on params/config.go:
```diff
@@ -364,91 +274,142 @@ type NodeConfig struct {
+		Version:         Version,
+		RPCEnabled:      RPCEnabledDefault,
+		HTTPHost:        HTTPHost,
+		HTTPPort:        HTTPPort,
```
The same.

adambabik commented on params/config.go:
```diff
@@ -364,91 +274,142 @@ type NodeConfig struct {
+		HTTPHost:        HTTPHost,
+		HTTPPort:        HTTPPort,
+		ListenAddr:      ListenAddr,
+		APIModules:      APIModules,
```
Same here. Generally, I would get rid of `defaults.go` completely.

pedropombeiro replied:
I removed about half the constants, but some are still shared with mailserver-canary and/or tests.

adambabik replied:
We can create a follow-up PR to remove `defaults.go` completely. In my opinion, it should not exist because the usage of this package spreads everywhere.

adambabik commented on params/config.go:
```diff
@@ -364,91 +274,142 @@ type NodeConfig struct {
-	if err != nil {
-		return nil, err
+// LoadConfigFromJSON parses incoming JSON and returned it as Config
+func LoadConfigFromJSON(configJSON string, nodeConfig *NodeConfig) error {
```
I am not sure if it's just be but if a method takes a pointer to an object and overrides it, I would call it `UpateConfigFromJSON(config *NodeConfig, configJSON string)`. Otherwise, if it starts with `Load`, I expect it to return a new object.

adambabik commented:
Also, there are some files that miss a new line at the end.

pedropombeiro commented:
@adambabik the only ones I see missing newlines are JSON files. Is that correct?

adambabik commented:
I tried to use it and there are a few things that I believe we should improve even though they might change how `statusd` acts by default.

First, let's run `statusd` and see the output:
```
$ ./build/bin/statusd
INFO [09-13|11:53:56.126] Status backend initialized
INFO [09-13|11:53:56.130] LES protocol is disabled                 package=status-go/node
INFO [09-13|11:53:56.130] SHH protocol is disabled                 package=status-go/node
INFO [09-13|11:53:56.130] Status service api is disabled           package=status-go/node
```

This is pretty useless node. Also, I don't see basic information like `datadir` and `networkid`.

Quite the same story continues when I used a config file:
```
$ ./build/bin/statusd -c static/config/cli/fleet-eth.staging.json
INFO [09-13|11:54:42.367] Status backend initialized
INFO [09-13|11:54:42.371] LES protocol is disabled                 package=status-go/node
INFO [09-13|11:54:42.371] SHH protocol is disabled                 package=status-go/node
INFO [09-13|11:54:42.371] Status service api is disabled           package=status-go/node
```

I am thinking about the following things:
1. Whisper is enabled by default,
2. Discovery is enabled by default when starting without config or using a fleet,
3. There is `-log` flag in CLI,
4. Can we move config JSON files to `config/` directory in the project root? It's pretty hidden now very hard to locate.

--- 

The whole comment is about CLI. When using status-react, whisper and discovery should be explicitly enabled.

pedropombeiro commented:
@adambabik OK, pushed commit with requested changes.

adambabik commented:
@PombeirP thanks! I though about it a bit different though. I took a look at `fleet-eth.beta.json` and:
- `NetworkID` -- I think it's unnecessary there. Network has nothing to do with a fleet,
- `DataDir`, `KeyStoreDir` as well,
- `NoDiscovery` should be `false` because we want to use the fleet and at the same time `Rendezvous` is set to `true` which will use rendezvous peers discovery.

`nodeConfig.NoDiscovery` override could be removed from `NewNodeConfigWithDefaults` then.

adambabik commented on params/config.go:
```diff
@@ -286,15 +286,16 @@ func NewNodeConfigWithDefaults(dataDir, fleet string, networkID uint64) (*NodeCo
+			err = LoadConfigFromJSON(string(statusConfigJSON), nodeConfig)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Default config could not be loaded: %s", err)
```
Errors should start with lower case.

pedropombeiro replied:
oh, missed that. Will fix it and merge, if no one else says anything against.

pedropombeiro replied:
wonder if there's a linter extension to catch that

divan commented:
> I don't see any relation between reading env variables and security.

Sure, there is none in this context. In mentioned _"security through obscurity"_ the keyword was "_obscurity_", not "_security_". :)
